CSharp/Library/QnAMaker/QnAMakerDialog.cs

```diff
@@ -1,110 +1,225 @@
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Dialogs;
+using Microsoft.Bot.Connector;
+using Microsoft.Bot.Builder.Internals.Fibers;
+using System.Reflection;
+using System.Linq;
+using System.Web;
+
+namespace Microsoft.Bot.Builder.CognitiveServices.QnAMaker
+{
+    /// <summary>
+    /// A dialog specialized to handle QnA response from QnA Maker.
+    /// </summary>
+    [Serializable]
+    public class QnAMakerDialog : IDialog<IMessageActivity>
+    {
+        protected readonly IQnAService[] services;
+        
+        public IQnAService[] MakeServicesFromAttributes()
+        {
+            var type = this.GetType();
+            var qnaModels = type.GetCustomAttributes<QnAMakerAttribute>(inherit: true);
+            return qnaModels.Select(m => new QnAMakerService(m)).Cast<IQnAService>().ToArray();
+        }
+
+        /// <summary>
+        /// Construct the QnA Service dialog.
+        /// </summary>
+        /// <param name="services">The QnA service.</param>
+        public QnAMakerDialog(params IQnAService[] services)
+        {
+            if (services.Length == 0)
+            {
+                services = MakeServicesFromAttributes();
+            }
+
+            SetField.NotNull(out this.services, nameof(services), services);
+        }
+
+        async Task IDialog<IMessageActivity>.StartAsync(IDialogContext context)
+        {
+            context.Wait(MessageReceivedAsync);
+        }
+
+        public async Task MessageReceivedAsync(IDialogContext context, IAwaitable<IMessageActivity> argument)
+        {
+            var message = await argument;
+            var qnaMakerResults = default(QnAMakerResults);
+
+            if (message != null && !string.IsNullOrEmpty(message.Text))
+            {
+                var tasks = this.services.Select(s => s.QueryServiceAsync(message.Text)).ToArray();
+
+                if (tasks.Any())
+                {
+                    var maxValue = tasks.Max(x => x.Result.Answers[0].Score);
+                    qnaMakerResults = tasks.First(x => x.Result.Answers[0].Score == maxValue).Result;
+
+                    var filteredQnAResults = new List<QnAMakerResult>();
+                    foreach (var qnaMakerResult in qnaMakerResults.Answers)
+                    {
+                        qnaMakerResult.Score /= 100;
+                        if (qnaMakerResult.Score >= qnaMakerResults.ServiceCfg.ScoreThreshold)
+                        {
+                            qnaMakerResult.Answer = HttpUtility.HtmlDecode(qnaMakerResult.Answer);
+                            filteredQnAResults.Add((qnaMakerResult));
+                        }
+                    }
+
+                    qnaMakerResults.Answers = filteredQnAResults;

```

Comment:
>filteredQnAResults [](start = 46, length = 18)

refactor the filter logic above and have it done automatically or make it an extension method for List<QnAMakeResult>